text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+### Git via HTTP
+
+This structure lends itself to be very easily be transfered using HTTP. Git knows at least four transport modes: Git, SSH, HTTP (dumb) and HTTP (smart). The dumb HTTP protocol requires an index to be created via `git update-server-info` (typically through a post-commit hook) but it works pretty much like one would expect:
+
+```console
+$ git update-server-info
+$ python3 -m http.server -b 127.0.0.1
+Serving HTTP on 127.0.0.1 port 8000 (http://127.0.0.1:8000/) ...
+```
+
+The `python3 -m http.server` is a simple HTTP server that ships with Python. With the server running we can, in a different terminal, try:
+
+```console
+$ git clone http://localhost:8000/.git
+Cloning into 'localhost'...
+```
+
+We can see the requests in the other terminal:
+
+```text
+127.0.0.1 - - [13/Aug/2020 22:47:31] "GET /info/refs?service=git-upload-pack HTTP/1.1" 404 -
+127.0.0.1 - - [13/Aug/2020 22:47:38] "GET /.git/info/refs?service=git-upload-pack HTTP/1.1" 200 -
+127.0.0.1 - - [13/Aug/2020 22:47:38] "GET /.git/HEAD HTTP/1.1" 200 -
+127.0.0.1 - - [13/Aug/2020 22:47:38] "GET /.git/objects/23/d9244c918044b4f1e28de295065b41e029f37c HTTP/1.1" 200 -
+127.0.0.1 - - [13/Aug/2020 22:47:38] "GET /.git/objects/5c/29792762ac54c0ce7614621c48f88eb5fb0352 HTTP/1.1" 200 -
+...
+```
+
+Since the structure is simple and we know the URL for every object if we know its hash, it is straightforward to fetch a single object from the server. Other transports are a bit more sophisticated, but the core idea is similar.
+
+### Pack files
+
+We do notice that our repo has quite a few objects:
+
+```console
+$ find .git/objects/ -type f | wc -l
+24
+```
+
+We can run garbage collection on these objects, causing git to remove unreachable ones and "pack" the others:
+
+```console
+$ git gc
+...
+$ find .git/objects/ -type f | wc -l
+4
+```
+
+Where did our files go? They are in a pack file:
+
+```console
+$ git verify-pack --verbose ./.git/objects/pack/pack-80a4d68ade4dcac5913a0e825ff50c6cb968c5a8.idx
+23d9244c918044b4f1e28de295065b41e029f37c commit 1122 848 12
+ff51a7fc40b7fb6f165d3a92ab88d8d2bf8c0b18 commit 1117 844 860
+...
+```
+
+Pack files are archives that contain multiple objects. They make things a little easier on the filesystem and the HTTP transfers, but more importantly delta compression is applied when files are in a pack.
+
+### Git summary
+
+Git stores different types of objects indexed by hashes, compressed in a very simple key-value scheme on the local filesystem. Every object contains a header indicating its type (commit, tree, blob). These objects can be fetched by hash even across the network. Parent relationships are indicated by hash as well.
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+* What is the actual impact on the execution engine?
+* Do we serialize hashes or contents?
+* Do we store DAG nodes statically or dynamically (`Box<dyn ...>`)?
+
+All these questions should be answered before this RFC is merged.
```

Comment:
> All these questions should be answered before this CEP is merged.

Point of order question: do you want us to discuss these items on this CEP or do you want to plan one or more discovery tasks?

Reply:
We will likely need more of these, one think I can think of is breaking out a separate CEP that examines how the consensus DAG (as opposed to the chain DAG) can be made to fit this scheme (and whether or not it should).

But I consider general support for the approach laid out here a prerequisite.